File: python/train_net.py
import os
import numpy as np
from scipy.io import loadmat
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import train_test_split
import joblib
from tkinter import Tk, filedialog
from helper_functions.MakeMiniMatFS import make_mini_mat_fs
from helper_functions.MakeTrainMat import make_train_mat
from parameters import *

signal = np.load(peak_file_path)
signal1 = np.load("helper_functions/SlowMini.npy")

noise_trace = np.load(noise_file_path)
noise_trace = noise_trace * 2.5

# Create training data

x = make_mini_mat_fs(N, SL, amp, mode, signal, signal1)[0]
X, y = make_train_mat(-x, noise_trace.flatten(), N, SL)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

# Initialize neural net
mlp = MLPClassifier(hidden_layer_sizes=(200, 100, 100), activation="logistic", max_iter=10000)

# Train the neural net
trainedNetN = mlp.fit(X, y)

# Save the trained model
joblib.dump(trainedNetN, result_network_path)

File: MATLAB/train_net.m
%% load the canonical peak shape

% set the correct working directory (same directory this file is located in) 
wd = matlab.desktop.editor.getActive;
wd = fileparts(wd.Filename);
cd(wd);
addpath 'helper_functions';

% prompt window to get the peak shape file. Should be a dx1 array stored in
% a .mat file. Check out 'FastMiniBo.mat' for the format of the signal
[peakFileName, peakFilePath] = uigetfile('*.mat');
load([peakFilePath, peakFileName]);
signal = mini;     % set signal to be the variable name that stores the canonical peak shape

load([peakFilePath, peakFileName]);


%% load the noise file

% prompt window to get the noise file
[noiseFileName, noiseFilePath] = uigetfile('*.mat');
load([noiseFilePath, noiseFileName]);
ASnoise = ASnoise * 2.5;                            % hyperparameter, scale the noise 

%% create training data

n = 10000;                          % training set contains n positives and n negatives
SL = size(signal, 1);               % sweeping length. number of data points for one peak
amp = 5;                            % hyperparameter, mean peak amplitude of the signals in the training set     

% [x] = buildTrainSet(n, SL, amp, 1, signal);     % create n signals with varying width and amplitude
[x] = MakeMiniMatFS(n, SL, amp, 1, signal);

% [X, Y]=MakeTrainMat(-x,reshape(ASnoise, [1 numel(ASnoise) ]), n, SL);% mini mean=StNset
[X, Y]=MakeTrainMat(-x,reshape(ASnoise, [1 numel(ASnoise) ]), n, SL);% mini mean=StNset

net0 = patternnet([200 100 100]);   % initialize neural net  
netN = configure(net0,X,Y);
[trainedNet,~] = train(netN,X,Y);

%%
uisave({'trainedNet'});

